### PR TITLE
Persist FxA/VPN attribution over multiple page navigations (Fixes #13147)

### LINKED ISSUE #13147: Persist FxA/VPN attribution over multiple page navigations
## Description

Right now our 1st party attribution setup for FxA/VPN simply passes through referral parameters and other data to FxA subscription links. It does not store or collect data for any length of time. It is also responsible for passing through coupon parameters so they can be automatically applied at checkout.

Now that we're aiming to move to a dedicated pricing page, which is separate to the main landing page, we're going to need to start storing that data for a user's session. This will allow people to navigate through several VPN related pages, whilst still ensuring their original attribution is retained when making a purchase.

Because we'll need to store some data in the user's browser, we need to think about an opt-out mechanism. My first thought is we could use the [first party data preferences](https://www.mozilla.org/en-US/privacy/websites/data-preferences/) page, although this would likely require some slight copy updates.

We should also ensure we always store/pass someone's coupon, if they visit via a coupon link. This should happen irrespective of their data preferences as it's related to site functionality.

---

#### Success Criteria
- [x] FxA/VPN attribution persists for a session/explicit length of time.

### PR DESCRIPTION
**Pull request by alexgibson**
## One-line summary

Updates FxA/VPN attribution to store session attribution data to persist over multiple page navigations.

## Significant changes and points to review

Note for reviewer: attribution parameters will now persist for the duration of your tab remaining open (using [sessionStorage](https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage)). So it's best to close / re-open the tab after each test below:

## Issue / Bugzilla link

#13147

## Testing

Single page view tests:

- [x] UTM parameters are passed through to subscription links: https://www-demo1.allizom.org/en-US/products/vpn/?utm_source=test&utm_campaign=test
- [x] Coupon parameters is passed through to subscription links: https://www-demo1.allizom.org/en-US/products/vpn/?coupon=test
- [x] UTM and coupons are passed through to subscription links: https://www-demo1.allizom.org/en-US/products/vpn/?utm_source=test&utm_campaign=test&coupon=test

Multiple page view tests:

1. Open https://www-demo1.allizom.org/en-US/?utm_source=test&utm_campaign=test
2. Scroll to the "Featured Mozilla Products" section and click "Get Mozilla VPN"

- [x] UTM parameters from step 1 are passed through to subscription links after page navigation

WNP test (clear cookies after this test):

1. Open https://www-demo1.allizom.org/en-GB/firefox/113.0/whatsnew/
2. Click "Get Mozilla VPN"

- [x] Verify `utm_campaign=whatsnew-113-eu` is passed through to subscription links.

Opt-out test (clear cookies after this test):

1. Open https://www-demo1.allizom.org/en-US/privacy/websites/data-preferences/
2. Click "Opt out of first-party data collection"
3. Close tab
4. Open https://www-demo1.allizom.org/en-US/products/vpn/?utm_source=test&utm_campaign=test
5. Verify the `utm_source=test&utm_campaign=test` params are **not** passed through to subscription links. Only the default set of UTM params should be present (`utm_source=www.mozilla.org-vpn-product-page` and `utm_campaign=vpn-product-page`).

Opt-out test with coupon (clear cookies after this test):

1. Open https://www-demo1.allizom.org/en-US/privacy/websites/data-preferences/
2. Click "Opt out of first-party data collection"
3. Close tab
4. Open https://www-demo1.allizom.org/en-US/products/vpn/?coupon=test
5. Verify coupon is still passed, even though opt-out was clicked.

### CODE REVIEW

**Comment by alexgibson** [comment]
Depends on https://github.com/mozilla/bedrock/issues/13165

**Comment by alexgibson** [comment]
This is still do-not-merge, but is ready for code review.

File: media/js/base/fxa-attribution.es6.js [hunk 1]
```diff
@@ -248,6 +265,52 @@ FxaAttribution.getCouponData = (params) => {
...
+    try {
+        return JSON.parse(window.sessionStorage.getItem(_mozReferralStorageID));
+    } catch (e) {
+        return false;
+    }
+};
+
+FxaAttribution.setReferralStorage = (data) => {
+    if (!data) {
+        return;
```
Comment (stephaniehobson) [inline media/js/base/fxa-attribution.es6.js:301]: Should this return false explicitly like the other failures?

File: media/js/base/fxa-attribution.es6.js [hunk 2]
```diff
@@ -319,38 +382,59 @@ FxaAttribution.appendToProductURL = (url, data) => {
...
+        if (referralData) {
+            params = referralData;
         } else {
-            // Check to for search engine referral data.
-            const searchData = FxaAttribution.getSearchReferralData();
+            const utmData = FxaAttribution.getUtmData(urlParams);
+            const experimentData = FxaAttribution.getExperimentData(urlParams);
+
+            // If there are UTM params in the page URL, then
+            // validate those as referral data.
```
Comment (stephaniehobson) [inline media/js/base/fxa-attribution.es6.js:406]: So someone with stored UTM info will have that over-ridden if they visit with new UTMs in the URL?

**Review by stephaniehobson — COMMENTED** [review]
All functionality is 💯. Love to see the tests.

I commented with a couple questions about how it's intended to work but don't see any needed changes.

File: media/js/base/fxa-attribution.es6.js (hunk 2 above)
Comment (alexgibson) [reply media/js/base/fxa-attribution.es6.js:406]: If someone has stored referral data, that will always be used (so their first entry page to the site). The data persists from when the active tab was opened to when it gets closed again.

It's possible that someone might enter twice, through two separate campaigns, during that tab's lifetime. In this case, their first entry point would be attributed.

Does that make sense?

**Comment by stephaniehobson** [comment]
Needs a rebase after https://github.com/mozilla/bedrock/pull/13298

**Comment by alexgibson** [comment]
If this ever returns it will require re-working given the upcoming consent changes. Going to close this as opposed to keep rebasing it.

**Pull request closed without merging**